File: python/02/second.py
def index_of_fault(levels):
    last_diff = None
    for i, l in enumerate(levels[:-1]):
        diff = levels[i + 1] - l
        if abs(diff) > 3 or diff == 0:
            return i
        if last_diff is not None:
            if diff * last_diff < 0:
                return i
        last_diff = diff
    return None


def is_safe(levels):
    i = index_of_fault(levels)

    # no removal ok?
    if i is None:
        # print(levels)
        return True

    # remove first ok?
    if i == 1:
        part = levels[i:]
        if index_of_fault(part) is None:
            print(i, i - 1, levels, part)
            return True

    # remove i ok?
    part = levels[:i] + levels[i + 1 :]
    if index_of_fault(part) is None:
        print(i, i, levels, part)
        return True

    # remove i + 1 ok?
    part = levels[: i + 1] + levels[i + 2 :]
    if index_of_fault(part) is None:
        print(i, i + 1, levels, part)
        return True
    return False


def solve(input):
    out = 0
    for row in input:
        if is_safe(list(row)):
            out += 1
    return out


def read(filename):
    with open(filename, "r") as f:
        rows = [row.rstrip() for row in f.readlines()]
        input = []
        for row in rows:
            row = row.split(" ")
            row = map(str.strip, row)
            row = map(int, row)

            input.append(row)
            # input.append(int(row))
            # input.append(list(row))
        return input


def main(filename):
    return solve(read(filename))


if __name__ == "__main__":
    import sys

    filename = "input.txt"
    # filename = "test.txt"

    if len(sys.argv) < 2:
        print("{}\n".format(main(filename)))
    else:
        for f in sys.argv[1:]:
            print("{}:\n{}\n".format(f, main(f)))

File: rust/day_02/second.rs
use std::fs;
use std::str::FromStr;

fn parse_input<T: FromStr>(contents: String) -> Vec<Result<Vec<T>, T::Err>> {
    let mut output = Vec::new();
    for row in contents.lines() {
        let parts = row.split_whitespace().map(|word| word.parse()).collect();
        output.push(parts);
    }
    return output;
}

fn is_safe<'a, I>(mut values: I) -> bool
where
    I: Iterator<Item = &'a i32>
{
    let mut last_diff = 0;
    let mut prev = values.next().unwrap();
    for val in values {
        let diff = prev - val;
        if diff == 0 || diff.abs() > 3 {
            return false;
        }
        if diff * last_diff < 0 {
            return false;  
        }
        last_diff = diff;
        prev = val;
    }
    return true;
}

fn solve(rows: Vec<Result<Vec<i32>, <i32 as FromStr>::Err>>) -> i32 {
    let mut output = 0;
    for row in rows {
        let r = row.unwrap();
        if is_safe(r.iter()) {
            output += 1;
        }
        else {
            for i in 0..r.len() {
                // try without the number at index i
                if is_safe(r.iter().enumerate().filter(|&(j, _)| j != i).map(|(_, e) | e)) {
                    output += 1;
                    break;
                }
            }
        }
    }
    return output;
}

fn main() {
    let file_name = "input.txt";
    // let file_name = "test.txt";
    let contents = fs::read_to_string(file_name).expect("Should have been able to read the file");
    let rows = parse_input::<i32>(contents);
    let answer = solve(rows);
    println!("{answer}");
}